### ISSUE TITLE
Training fails for method 'gpu_hist' for large datasets with missing values

### ISSUE THREAD

**Issue by tpboudreau** [dmlc/xgboost#7025]
When a training dataset is sparse (contains missing values) and the number of elements in the dmatrix exceeds 2^32, attempts to train using tree_method=gpu_hist can fail with a diverging RMSE.

For example, using the dataset [here](https://www.gamma-rho.com/xgboost/divergent_training_dataset.csv.gz) -- which is 2 million rows by 2500 features ( = 5x10^9 > 2^32 elements ) -- and this code:

```
import xgboost as xg

if __name__ == '__main__':

    dm = xg.DMatrix(data='divergent_training_dataset.csv?format=csv&label_column=0')
    print("training matrix: {}x{}".format(dm.num_row(), dm.num_col()))

    pm = {
      'objective': 'reg:squarederror',
      'max_depth': 3,
      'tree_method': 'gpu_hist',
      'sampling_method': 'uniform',
      'subsample': 0.3,
      'seed': 121,
    }
    print(pm)

    md = xg.train(params=pm, dtrain=dm, num_boost_round=25, evals=[(dm, "training")])

```
produces the following result:

> training dmatrix: 2000000x2500
> {'objective': 'reg:squarederror', 'max_depth': 3, 'tree_method': 'gpu_hist', 'sampling_method': 'uniform', 'subsample': 0.3, 'seed': 121}
> [0]     training-rmse:2033.44800
> [1]     training-rmse:892402.00000
> [2]     training-rmse:39885900.00000
> [3]     training-rmse:154548464.00000
> [4]     training-rmse:6840178176.00000
> [5]     training-rmse:51539161088.00000
> [6]     training-rmse:443111047168.00000
> [7]     training-rmse:10977591427072.00000
> [8]     training-rmse:10590034591744.00000
> [9]     training-rmse:10473789456384.00000
> [10]    training-rmse:10343798538240.00000
> [11]    training-rmse:10533048680448.00000
> [12]    training-rmse:6130236915712.00000
> [13]    training-rmse:87665089183744.00000
> [14]    training-rmse:1312450603384832.00000
> [15]    training-rmse:inf
> [16]    training-rmse:inf
> [17]    training-rmse:inf
> [18]    training-rmse:inf
> [19]    training-rmse:inf
> [20]    training-rmse:inf
> [21]    training-rmse:inf
> [22]    training-rmse:inf
> [23]    training-rmse:inf
> [24]    training-rmse:inf

_NOTE: the linked dataset is ~15GB compressed and ~86GB uncompressed; the script requires 80-90GB CPU memory._

___

The root cause of the failure is a narrowing conversion when passing arguments to the binary search function
responsible for mapping dmatrix elements to ellpack page locations.

A patch is forthcoming.

**Comment by trivialfis** [comment]
Wow!  Thanks for submitting the issue and the fix!  It's exciting to see users getting larger and larger datasets.
